Filter generation

tfcollins opened the pull request:
Adds support for FIR filter generation.

rgetz commented:
shouldn't this be integrated into ad9361_baseband_auto_rate.c? If the filter designer exists - it should use that rather than the fixed FIRs that are there?

OR is that part of the #3 ?

tfcollins commented:
We need the top level function ad9361_generate_fir_taps.c for generic filter designs, but yes #3 should basically replace ad9361_baseband_auto_rate.c

tfcollins commented:
@rgetz do you want me to include a cook_inputs like function in this PR?

rgetz commented:
replace or augment?

I think we would want to keep all those function names the same, and just replace the inner workings to use the filter designer.

If it is in here, or a different one - is up to you or Michael.

-Robin

tfcollins commented:
Just need to finish up some test code and a cook_input function will be added to the existing functionality for the baseband sample rate function.

By default it will use the new designer, but without it the existing filter set will be used.

mhennerich commented:
The filter designer takes a lot of arguments. Which ones are you going to expose?

tfcollins commented:
@mhennerich I want to expose the general filter designer so filters can be generated to user specifications, but I can also provide a function that all you would need to supply is the sample rate and RX or TX.  This is basically the API MathWorks uses with the current designer based in MATLAB.

I can include that extra function in this PR if you want.

tfcollins commented:
@rgetz With these additional changes this should be a drop in replacement for osc.  baseband_auto_rate will use the filter designer when built into the library.

Calling "cmake -DFilterDesigner=OFF .." will give you the older functionality if needed.

mhennerich commented:
Design a filter only by the baseband rate? In order to compensate roll-off on the analog filters shouldn't we also use RF Bandwidth. Or better use RF Bandwidth to calculate the cut-off?

tfcollins commented:
So we do calculate an RF bandwidth for you (it's a guess but should not affect a signal at a max bandwidth of the data rate supplied), but its used to generate the compensation filter for both TX and RX filter chains.  So that aspect is taken into account.

Michael, would you want the user to supply this value? This would change the API to the baseband_auto_rate function.  Unless we wanted another automatic filter generation function.

mhennerich commented:
We definitely don't want to break existing API.
However we can add new derivatives anytime. I see some value in calculating the Filter based on BB rate and RF BW.

tfcollins commented:
@rgetz after travis-ci and appveyor are ready I'll do a merge.  Things seem to be working well with OSC for FMComms2 and the designs are pretty fast.

mhennerich commented:
1. I dislike the fact that the old flat response behavior is ultimately gone as soon as the lib is built with the filter designer. Ideally we create a new function that uses the filter designer if available otherwise return error.
2. Please check return values and use consistent coding style.
3. There are also some errors:
  * Taps return always 0 on the TX side (ad9361_generate_fir_taps(&fdp, firTX, &taps))
   * If rate > 40MSPS the taps number from the second call (ad9361_generate_fir_taps(&fdp, firRX, &taps)) will return 1, which causes an error.

mhennerich commented on ad9361_baseband_auto_rate.c:
```diff
@@ -82,9 +82,28 @@ int ad9361_set_bb_rate(struct iio_device *dev, unsigned long rate)
+	int16_t *firRX = (int16_t *)malloc(sizeof(int16_t)*128);
+	ad9361_generate_fir_taps(&fdp, firRX, &taps);
+	dec = (int)fdp.FIR;
+	chan = iio_device_find_channel(dev, "voltage0", true);
```
please check return value

mhennerich commented:
Also RF BW for RX and TX are calculated differently - however the code write the same value to both Rx and Tx.

rgetz commented:
As for 1 - it will never be "flat", since you don't know what the RF bandwidth is, and delta sigma will always push out noise to the outside.

mhennerich commented:
I agree - but these filters are pushing the passband to the limit. Basically the ideal filter if you want to do a wide spectrum view. Using the filter designer it doesn't allow you to design such a filter.

tfcollins commented:
This is just the default behavior of the filter designer as it exists in MATLAB.  Michael, would requiring an RF bandwidth argument fix this?  That would definitely give you a wider receive perspective if desired.

rgetz commented:
I don't think Michael was talking about RF bandwidth - more about pass band. 

If someone asks for x MSPS - by default - should you have any cutoff? The existing filters are just flat (all band pass).

tfcollins commented:
By default this is how we determine the Fpass and Fstop positions:
input.Fpass = input.Rdata / 3;
input.Fstop = input.Fpass * 1.25;

https://github.com/analogdevicesinc/ad936x-filter-wizard/blob/ca68f7d0b041f59fd117a7f55ac05b7f1edfd8ff/cook_input.m#L147

Would we want this changed so Fpass==Rdata?

rgetz commented:
according to me - it's fine (the delta sigma architecture pushes all the noise into the outside, and you want to throw that away anyway - sooner is better).

according to Michael - it's a change in the way the library worked from the past, so yes.

Since he clicks on the merge pull request button, that's fine with me. :) As long as there is a way to override that in the details. I can call cook inputs (equivalent), and then change the structure the way I want it (pull the pass band in), and then design the filter... As long as I can do that - it should be fine.

tfcollins commented:
Michael, I found a bug/typo on the filter wizard MATLAB code which should fix those errors you are seeing.  I need to finish up the test code to exercise the HB different permutations to avoid such bugs in the future.  Then I can update the generate code here.

tfcollins commented:
Michael, as per our discussion I just created a new branch for this work which does not touch the ad9361_baseband_auto_rate.

I've addressed both errors you mentioned and added some test code in the new branch and in the MATLAB code to cover these as well.  I've also updated the error checking and switched to use astyle.

We can talk offline with regards to an additional function to actually write filters to devices.  Right now I do nothing with the taps and additional configurations.  Maybe we can do something like libiio's tools.

mhennerich commented:
ok - let me take a crack at it.

mhennerich commented:
Hi,

I committed some WIP changes with 5a84bc838133ce3c940b36d95350bc0508eeeb62

However there seems to be a few problems:

Everything above 7MSPS will only result in FIR decimation of 1. (See [1] amd [2])
Also DAC div seems to be computed incorrectly (See [3]) - there is no reason to use DAC div here.
If you compare the rates computed by the designer with the ones the driver would compute there is a huge difference.

[1]
RX 3 GAIN 0 DEC 1
TX 3 GAIN 0 INT 1
RRX 1024000000 64000000 32000000 16000000 8000000 8000000
RTX 1024000000 64000000 32000000 16000000 8000000 8000000
BWRX 6455167
BWTX 6418490
38,40
39,41
-53,-57
[snip]

[2]
RX 3 GAIN -6 DEC 2
TX 3 GAIN 0 INT 2
RRX 896000000 112000000 56000000 28000000 14000000 7000000
RTX 896000000 112000000 56000000 28000000 14000000 7000000
BWRX 5648271
BWTX 5616179
0,-18
-1,-13
-4,23
[snip]

[3]
RX 3 GAIN 0 DEC 1
TX 3 GAIN 0 INT 1
RRX 960000000 240000000 120000000 60000000 30000000 30000000
RTX 960000000 120000000 60000000 30000000 30000000 30000000
BWRX 21613283
BWTX 22063560
43,40
43,41
-64,-57


From: Travis Collins [mailto:notifications@github.com]
Sent: Freitag, 6. Oktober 2017 03:06
To: analogdevicesinc/libad9361-iio <libad9361-iio@noreply.github.com>
Cc: Hennerich, Michael <Michael.Hennerich@analog.com>; Mention <mention@noreply.github.com>
Subject: Re: [analogdevicesinc/libad9361-iio] Filter generation (#4)


Michael, as per our discussion I just created a new branch for this work which does not touch the ad9361_baseband_auto_rate.

I've addressed both errors you mentioned and added some test code in the new branch and in the MATLAB code to cover these as well. I've also updated the error checking and switched to use astyle.

We can talk offline with regards to an additional function to actually write filters to devices. Right now I do nothing with the taps and additional configurations. Maybe we can do something like libiio's tools.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/analogdevicesinc/libad9361-iio/pull/4#issuecomment-334632608>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABFwob474HCa6J8vtdbHgoGg7T3-ye7Hks5spXz1gaJpZM4PmFj3>.

rgetz commented:
Michael:

Where in the driver does it calculate the rates? I thought we were doing it the same - but it could have diverged. We can manually check the algorithm.

mhennerich commented:
Please see here:
https://github.com/analogdevicesinc/linux/blob/xcomm_zynq/drivers/iio/adc/ad9361.c#L3992

I think the problem with the filter designer code is that it looks at RX and TX individually.
However in order to calculate the best plan you need to take both into account.

tfcollins commented:
The current designer really only enables the FIR decimation when it has too, and DAC_div by default are different between TX and RX.  Not sure where this came from.

Is there a specific strategy for selecting decimation other than run the ADC/DAC as fast as we can?  Do we want to use the FIR decimation even if we don't have to?

mhennerich commented:
Without the FIR interpolation - you can't max your oversampling ratio.

The higher the oversampling ratio, the higher the processing gain and SNR.
The better your SNR - the better the EVM.

The limits I've chosen in the ad9361_set_bb_rate() function are at the high end of what's possible.
(trying to run the ADC always as close to 640MSPS as possible)

https://github.com/analogdevicesinc/libad9361-iio/blob/master/ad9361_baseband_auto_rate.c#L88

Not a 100% sure if this makes too much sense or not.

tfcollins commented:
Yes I totally agree on the reasoning for setting things up at these rates.

I believe the only concern that I had was with regards to the chip speed limitations and how decimation was used and where.  The MATLAB function (and the template generator I wrote) explicitly check against the MIN/MAX rates of each HB section.  https://github.com/analogdevicesinc/ad936x-filter-wizard/blob/master/cook_input.m#L206

The "ad9361_calculate_rf_clock_chain" function may do this inherently but I wasn't obvious to me.

But yes I totally agree we need to max out the rates the best we can with the most decimation.
